Batched sample registration form with Paste capabilities

ramonski opened the pull request:
## Description of the issue/feature this PR addresses

This PR enhances the sample add form for better usability and data management

## Current behavior before PR

- All sample columns are displayed
- Only possible to copy the value of the first column over to the other columns

<img width="952" alt="Old Sample Add Form" src="https://github.com/user-attachments/assets/66cc5007-49b8-4b75-8eda-a2dd8759cfb7">


## Desired behavior after PR is merged

- Tabbed sample columns that allow individual navigation
- Paste window to enter individual values for sample columns


<img width="958" alt="New Sample Add Form" src="https://github.com/user-attachments/assets/662071d4-7adf-43de-8477-837730b13e5a">


--
I confirm I have tested this PR thoroughly and coded it according to [PEP8][1]
and [Plone's Python styleguide][2] standards.

[1]: https://www.python.org/dev/peps/pep-0008
[2]: https://docs.plone.org/develop/styleguide/python.html

ramonski committed:
Implemented deferred bounce method

ramonski committed:
Refactoring

ramonski committed:
Allow to set object/array values

ramonski committed:
Paste action

ramonski committed:
Implemented paste handler

ramonski committed:
Conditional copy/paste buttons

ramonski committed:
Tabbed sample add form

ramonski committed:
SVG added

ramonski committed:
Changed to fontwawesome icons

ramonski committed:
Handle service columns

ramonski committed:
Changelog updated

ramonski committed:
Added condition for manage form fields

ramonski committed:
Fixed multi reference field copy

ramonski committed:
Error handling for invalid dates

ramonski commented:
Something that might be confusing with the tabbed view is the "copy" button behavior.

It always uses the value of the first column (Sample 1) and copy it over to the other columns.

When you view now e.g. "Sample 3" and click this button, one might expect that the current displayed value is copied over to the other columns on the right, but that is not the case.

Hence, we might think of either hiding this `>>` button when switching to any other tab than "Sample 1" or adapt the functionality to copy always the value of the current displayed tab to the other columns.

What do you think?

xispa commented:
> Something that might be confusing with the tabbed view is the "copy" button behavior.
> 
> It always uses the value of the first column (Sample 1) and copy it over to the other columns.
> 
> When you view now e.g. "Sample 3" and click this button, one might expect that the current displayed value is copied over to the other columns on the right, but that is not the case.
> 
> Hence, we might think of either hiding this `>>` button when switching to any other tab than "Sample 1" or adapt the functionality to copy always the value of the current displayed tab to the other columns.
> 
> What do you think?

If it is not too difficult, I would adapt the functionality to always copy the value displayed in current tab. I think is the less confusing approach, yet the most powerful. Would hide the button when the current tab is not the first otherwise.

If you go ahead with adapting the current functionality, when the button is clicked and current tab is not the first, I would also display a confirmation pop-up:

*The system will copy the value '`<field_value>`' from field '`<field_name>`' across the rest of samples. 
Are you sure?
[Yes] [No]*

ramonski commented:
Thanks for your feedback Jordi!
IMO a confirmation dialog is only necessary when the functionality is not intuitive and logical and might get annoying otherwise.
Therefore, I would like to implement the copy functionality for the current tab only to all columns on the right. I guess this should be possible because the tab switching and the copy functionality are implemented in JS.

ramonski commented:
Of course I would change the title attribute of the copy button for an updated description. But I guess the new button icon (`>>`) should hint enough what the functionality is doing, aka. copy value to the right.

ramonski committed:
Copy to columns on the right

ramonski committed:
Merge branch '2.x' into ar-add-paste-field-values

ramonski commented:
Ok, implemented the "copy to columns on the right" feature and the PR is ready for review again

xispa commented:
Man this is coool!. Haven't reviewed in detail yet, but I wanted to share a concern about the overlay for the introduction of individual values all-at-once:

![Captura de 2024-12-18 16-14-46](https://github.com/user-attachments/assets/5e2a6502-267b-4ddb-8527-7205cafb0daf)

This component is not specific of field type. User can manually type the text value for each sample. This works great for text-based field, but I think is prone to errors for fields from other types (e.g. select, reference, checkbox, etc.). For instance, if the field is a reference field and the text does not match any value, the system simply flushes the value of the field. And this might be a problem, specially taking into account that form values would usually remain "hidden" under their own tabs. Other examples in wihch this component might lead to confusion are `select` fields or custom-widget fields like patient's MRN or Date of birth.

Therefore I think would be better to display the icon for the introduction of individual values all-at-once only for text-based fields. However, I presume that some labs might want this functionality to be available even for fields not fully-compatible with the type (e.g. datetime), maybe because they do a copy&paste from a spreadsheet or similar that comes with the right format from the beginning. So an alternative would be to rely on a senaite registry setting to either render the icon or not for a given sample field.

What do you think?

ramonski commented:
Yes, this paste option might get tricky with these custom fields. Currently, this code handles it: https://github.com/senaite/senaite.core/pull/2658/files#diff-e3237269d12664a4fc5c68bf03f6cc8b5ef8cd8a82b36168beb3d48ed05792ffR983-R1018

```coffee
paste_value: (field, value) =>
    # handle reference field
    controller = @get_widget_controller(field)
    if controller
      # NOTE: We use here the search API to find the entered text and set it
      # only when we find exactly one result
      promise = controller.search value
      promise.then (data) ->
        # only set the value if we get exactly 1 search result
        if data["count"] != 1
          return
        # select the result
        items = data["items"]
        controller.select(items)
        controller.clear_results()
    # set all other fields
    else if @is_text(field) or @is_textarea(field)
      field.value = value
    else if @is_radio(field) or @is_checkbox(field)
      try
        checked = Boolean(JSON.parse(value))
      catch
        checked = no
      field.checked = checked
    # date field
    else if @is_date_widget(field)
      # we need to fetch the date and time input fields
      date_input = field.querySelector("input[type='date']")
      time_input = field.querySelector("input[type='time']")
      try
        date = new Date(value)
        date_input.value = date.toISOString().split("T")[0]
        time_input.value = date.toTimeString().split(" ")[0]
      catch error
        console.warn error
        site.add_notification("Invalid date format", "Please use the format yyyy-mm-dd MM:HH")
```

This means that only exact matches for all reference fields are used, otherwise, the field is skipped.

Besides this, Checkbox, Text and Date fields can be (somehow) handled.

Maybe you can give it a try with the Patient fields and see what side-effects happen. If it get's too weird, we can maybe disable the button based on a registry setting...

ramonski committed:
Merge branch '2.x' into ar-add-paste-field-values

ramonski committed:
Merged 2.x

ramonski committed:
Production JS built

ramonski committed:
Merge branch 'ar-add-paste-field-values' of https://github.com/senaite/senaite.core into ar-add-paste-field-values

ramonski committed:
Add registry record to control multi paste button appearance

ramonski committed:
Description updated

ramonski commented:
I've added a Registry setting that allows to specify the field names that allow multi-pasting.
This field is exclusive, that means that if one value is set, all values must be set for allowed fields. Per default, only string, boolean, reference and date fields are supported.

xispa committed:
Merge branch '2.x' into ar-add-paste-field-values

xispa commented on src/bika/lims/browser/analysisrequest/add2.py:
```diff
@@ -563,6 +572,42 @@ def is_service_selected(self, service):
+            widget_type = None
+        if widget_type not in ALLOW_MULTI_PASTE_WIDGET_TYPES:
+            return False
+        return True
```
I would simplify with `return widget_type in ALLOW_MULTI_PAST_WIDGET_TYPES`

xispa commented on src/bika/lims/browser/analysisrequest/templates/ar_add2.pt:
```diff
@@ -412,16 +434,36 @@
+                       title="Copy the value of the first sample to the others"
+                       data_toggle="tooltip"
+                       tal:attributes="fieldName python:fieldName;
+                                       fieldLabel python:fieldLabel">
```
`i18n:attributes="title"` is missing

xispa commented on src/bika/lims/browser/analysisrequest/templates/ar_add2.pt:
```diff
@@ -412,16 +434,36 @@
+                       data_toggle="tooltip"
+                       tal:condition="python:view.show_paste_button_for(field)"
+                       tal:attributes="fieldName python:fieldName;
+                                       fieldLabel python:fieldLabel">
```
`i18n:attributes="title"` is missing

xispa commented on src/bika/lims/browser/analysisrequest/templates/ar_add2.pt:
```diff
@@ -516,20 +558,27 @@
+                          <!-- Copy Button -->
+                          <a class="copy d-block"
+                             title="Copy the currently displayed column value to the other columns on the right"
+                             data_toggle="tooltip">
```
`i18n:attributes="title"` is missing

xispa commented on src/senaite/core/browser/static/js/coffee/bika.lims.analysisrequest.add.coffee:
```diff
@@ -951,6 +957,69 @@ class window.AnalysisRequestAdd
+        time_input.value = date.toTimeString().split(" ")[0]
+      catch error
+        console.warn error
+        site.add_notification("Invalid date format", "Please use the format yyyy-mm-dd MM:HH")
```
Not sure if I am pointing to the wrong place, but I've noticed that when the value pasted has the `dd/mm/yyyy` format, the TZ is not taken into account. See this use case:

![Captura de 2025-02-07 10-16-06](https://github.com/user-attachments/assets/53470dd6-aa9b-4da9-bea2-3c6d7460551e)

And when I click to "Paste", the value set to the field is "31-12-2024":

![Captura de 2025-02-07 10-16-14](https://github.com/user-attachments/assets/f75e5e5b-85b7-467e-9475-3709cb21879e)

However, when the value to be pasted has the `yyyy-mm-dd` format:

![Captura de 2025-02-07 10-21-09](https://github.com/user-attachments/assets/17cb818c-1e4f-4487-b777-c64f6e146563)

The TZ is preserved:

![Captura de 2025-02-07 10-21-18](https://github.com/user-attachments/assets/df1dc066-5d3f-466d-977b-2639574c26a4)

Well... seems the pasting when the format is `dd/mm/yyyy` does not work as it should. If I try to paste `02/02/2025`:

![Captura de 2025-02-07 10-26-00](https://github.com/user-attachments/assets/c6e57cb0-958f-4a9d-a7f8-1baaa4dc1e44)

I end up this:

![Captura de 2025-02-07 10-26-08](https://github.com/user-attachments/assets/5528a50c-77c2-476d-b767-96e017ed2b80)


This said,  wouldn't I expect to see a notification with "Invalid date format" when the format is not `yyyy-mm-dd` though?

xispa replied:
And sometimes seems to work, but mixes the day with the month:

![Captura de 2025-02-07 10-32-51](https://github.com/user-attachments/assets/393b6027-3f44-4c1b-b7da-9995e96b4639)
![Captura de 2025-02-07 10-32-59](https://github.com/user-attachments/assets/751198ee-1ed5-437d-83ef-8e038154887b)

ramonski replied:
Yes, the only compatible format is YYYY-mm-dd HH:MM
I'm going to disable date pasting for per default

ramonski replied:
> Not sure if I am pointing to the wrong place, but I've noticed that when the value pasted has the `dd/mm/yyyy` format, the TZ is not taken into account

Yes, the timezone is never set in the Widget HTML, but is done later in the Python logic. Therfore, we always handle timezone naive date and time strings.

However, since we allow now to set the TZ per user, we should actually use their format instead of the server set TZ:
<img width="405" alt="User TZ" src="https://github.com/user-attachments/assets/4de36131-470f-4a73-a9b1-03b8c111ef73" />

ramonski replied:
Disabled multi paste for date fields in [15231ed](https://github.com/senaite/senaite.core/pull/2658/commits/15231ed55958e3922ca716dcd62f5304830cea07)

xispa reviewed (changes requested):
Awesome functionality!. However, it does not seem to work properly when pasting dates (see comments).

A part from that, I was hitting the "Paste" button from the "Multipaste popup" but I was not seeing any change in the field... I was puzzled until I realized I was located in Sample 2, but only had a value set in the multipaste popup for Sample 1.... 

Would be possible to highlight or something the number of sample you are actually looking at when you are in the "Multipaste popup"? Something like the following, so user do knows that is looking at "Sample 2":

![Captura de 2025-02-07 10-43-45](https://github.com/user-attachments/assets/c4ab8926-4fc5-4dba-ba59-47210b99db22)

Alternatively, warn the user when hitting "Paste" but the value for the current sample is not set. This would reduce the chance of errors.

ramonski committed:
Highlight the line of the current visible tab

ramonski committed:
Merge branch 'ar-add-paste-field-values' of https://github.com/senaite/senaite.core into ar-add-paste-field-values

ramonski commented:
> Awesome functionality!. However, it does not seem to work properly when pasting dates (see comments).
> 
> A part from that, I was hitting the "Paste" button from the "Multipaste popup" but I was not seeing any change in the field... I was puzzled until I realized I was located in Sample 2, but only had a value set in the multipaste popup for Sample 1....
> 
> Would be possible to highlight or something the number of sample you are actually looking at when you are in the "Multipaste popup"? Something like the following, so user do knows that is looking at "Sample 2":
> 
> ![Captura de 2025-02-07 10-43-45](https://private-user-images.githubusercontent.com/832627/410856755-c4ab8926-4fc5-4dba-ba59-47210b99db22.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzg5NDE3MzMsIm5iZiI6MTczODk0MTQzMywicGF0aCI6Ii84MzI2MjcvNDEwODU2NzU1LWM0YWI4OTI2LTRmYzUtNGRiYS1iYTU5LTQ3MjEwYjk5ZGIyMi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMjA3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDIwN1QxNTE3MTNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jZGIyMjQ2MDQ5ZTE5ZDE2MWNjNTU4MWQ0ZWZmZmE5NTgxM2UwY2MwYTI1NWQxOTljZjg2YTVhNTc3ZDRmYjNiJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.q6QkiVbLS-WYivr3fqdeyt7T8H6RCdz4m2GmVgNbEWQ)
> 
> Alternatively, warn the user when hitting "Paste" but the value for the current sample is not set. This would reduce the chance of errors.

Addressed in [619ad0e](https://github.com/senaite/senaite.core/pull/2658/commits/619ad0ed8ceb56befe1131b679c8f7629f4bf156)

ramonski committed:
disable paste for date fields

ramonski committed:
Simplify

ramonski committed:
Unhighlight line for show-all

ramonski committed:
Added missing i18n attributes

ramonski committed:
Destroy the template dialog after close

xispa committed:
Merge branch '2.x' into ar-add-paste-field-values

xispa reviewed (approved):
Excellent, thanks!
